### ISSUE TITLE
nn.parallel.DistributedDataParallel nccl backend multiple-gpu multiple-node deadlock

### ISSUE THREAD

**Issue by jefby** [pytorch/pytorch#22531]
## 🐛 Bug

I want to use nn.parallel.DistributedDataParallell to train a model, its topo is 
- 30.4.241.186 GTX1080
- 30.4.241.12 GTX1080+GTX2080Ti


## To Reproduce

```python
from __future__ import division, print_function

import argparse

import torch
import torch.nn.functional as F
from torch import distributed, nn
from torch.utils import data
from torchvision import datasets, transforms


def distributed_is_initialized():
    if distributed.is_available():
        if distributed.is_initialized():
            return True
    return False


class Average(object):

    def __init__(self):
        self.sum = 0
        self.count = 0

    def __str__(self):
        return '{:.6f}'.format(self.average)

    @property
    def average(self):
        return self.sum / self.count

    def update(self, value, number):
        self.sum += value * number
        self.count += number


class Accuracy(object):

    def __init__(self):
        self.correct = 0
        self.count = 0

    def __str__(self):
        return '{:.2f}%'.format(self.accuracy * 100)

    @property
    def accuracy(self):
        return self.correct / self.count

    def update(self, output, target):
        with torch.no_grad():
            pred = output.argmax(dim=1)
            correct = pred.eq(target).sum().item()

        self.correct += correct
        self.count += output.size(0)


class Trainer(object):

    def __init__(self, model, optimizer, train_loader, test_loader, device):
        self.model = model
        self.optimizer = optimizer
        self.train_loader = train_loader
        self.test_loader = test_loader
        self.device = device

    def fit(self, epochs):
        for epoch in range(1, epochs + 1):
            train_loss, train_acc = self.train()
            test_loss, test_acc = self.evaluate()

            print(
                'Epoch: {}/{},'.format(epoch, epochs),
                'train loss: {}, train acc: {},'.format(train_loss, train_acc),
                'test loss: {}, test acc: {}.'.format(test_loss, test_acc),
            )

    def train(self):
        self.model.train()

        train_loss = Average()
        train_acc = Accuracy()

        for data, target in self.train_loader:
            data = data.to(self.device)
            target = target.to(self.device)

            output = self.model(data)
            loss = F.cross_entropy(output, target)

            self.optimizer.zero_grad()
            loss.backward()
            self.optimizer.step()

            train_loss.update(loss.item(), data.size(0))
            train_acc.update(output, target)

        return train_loss, train_acc

    def evaluate(self):
        self.model.eval()

        test_loss = Average()
        test_acc = Accuracy()

        with torch.no_grad():
            for data, target in self.test_loader:
                data = data.to(self.device)
                target = target.to(self.device)

                output = self.model(data)
                loss = F.cross_entropy(output, target)

                test_loss.update(loss.item(), data.size(0))
                test_acc.update(output, target)

        return test_loss, test_acc


class Net(nn.Module):

    def __init__(self):
        super(Net, self).__init__()
        self.fc = nn.Linear(784, 10)

    def forward(self, x):
        return self.fc(x.view(x.size(0), -1))


class MNISTDataLoader(data.DataLoader):

    def __init__(self, root, batch_size, train=True):
        transform = transforms.Compose([
            transforms.ToTensor(),
            transforms.Normalize((0.1307,), (0.3081,)),
        ])

        dataset = datasets.MNIST(root, train=train, transform=transform, download=True)
        sampler = None
        if train and distributed_is_initialized():
            sampler = data.DistributedSampler(dataset)

        super(MNISTDataLoader, self).__init__(
            dataset,
            batch_size=batch_size,
            shuffle=(sampler is None),
            sampler=sampler,
        )


def run(args):
    device = torch.device('cuda' if torch.cuda.is_available() and not args.no_cuda else 'cpu')

    model = Net()
    if distributed_is_initialized():
        model.to(device)
        model = nn.parallel.DistributedDataParallel(model)
        print("distributed_is_initialized")
    else:
        model = nn.DataParallel(model)
        model.to(device)

    optimizer = torch.optim.Adam(model.parameters(), lr=args.learning_rate)

    train_loader = MNISTDataLoader(args.root, args.batch_size, train=True)
    test_loader = MNISTDataLoader(args.root, args.batch_size, train=False)

    trainer = Trainer(model, optimizer, train_loader, test_loader, device)
    trainer.fit(args.epochs)


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument('--backend', type=str, default='gloo', help='Name of the backend to use.')
    parser.add_argument(
        '-i',
        '--init-method',
        type=str,
        default='tcp://127.0.0.1:23456',
        help='URL specifying how to initialize the package.')
    parser.add_argument('-s', '--world-size', type=int, default=1, help='Number of processes participating in the job.')
    parser.add_argument('-r', '--rank', type=int, default=0, help='Rank of the current process.')
    parser.add_argument('--epochs', type=int, default=20)
    parser.add_argument('--no-cuda', action='store_true')
    parser.add_argument('-lr', '--learning-rate', type=float, default=1e-3)
    parser.add_argument('--root', type=str, default='data')
    parser.add_argument('--batch-size', type=int, default=128)
    args = parser.parse_args()
    print(args)

    if args.world_size > 1:
        distributed.init_process_group(
            backend=args.backend,
            init_method=args.init_method,
            world_size=args.world_size,
            rank=args.rank,
        )

    import time
    start = time.time()
    run(args)
    end = time.time()
    print("it takes {} s".format(end - start))


if __name__ == '__main__':
    main()
```

on rank 0, ip address is 30.4.241.186, run command:

```bash
python main.py --init-method tcp://30.4.241.186:23456 --rank 0 --world-size 2 --backend nccl
```
on rank 1, ip address is 30.4.241.12, run command 

```bash
python main.py --init-method tcp://30.4.241.186:23456 --rank 1 --world-size 2 --backend nccl
```
it will enter deadlock like below:
rank 0
```
ZBOX-EN1080K:10379:10379 [0] NCCL INFO NET/Socket : Using [0]eth0:30.4.241.186<0>
ZBOX-EN1080K:10379:10379 [0] NCCL INFO NET/Plugin : No plugin found (libnccl-net.so).

ZBOX-EN1080K:10379:10379 [0] misc/ibvwrap.cu:63 NCCL WARN Failed to open libibverbs.so[.1]
NCCL version 2.4.2+cuda10.0
ZBOX-EN1080K:10379:10390 [0] NCCL INFO Setting affinity for GPU 0 to ff
ZBOX-EN1080K:10379:10390 [0] NCCL INFO comm 0x7ff1c0001920 rank 0 nranks 2 cudaDev 0 nvmlDev 0
ZBOX-EN1080K:10379:10390 [0] NCCL INFO CUDA Dev 0[0], Socket NIC distance :  PHB
ZBOX-EN1080K:10379:10390 [0] NCCL INFO Channel 00 :    0   1
ZBOX-EN1080K:10379:10390 [0] NCCL INFO NET/Socket : GPU Direct RDMA Disabled for GPU 0[0] / HCA 0 (distance 2 >= 2)
ZBOX-EN1080K:10379:10390 [0] NCCL INFO Ring 00 : 1 -> 0 [receive] via NET/Socket/0
ZBOX-EN1080K:10379:10390 [0] NCCL INFO Ring 00 : 0 -> 1 [send] via NET/Socket/0
ZBOX-EN1080K:10379:10390 [0] NCCL INFO Using 256 threads, Min Comp Cap 6, Trees disabled
ZBOX-EN1080K:10379:10390 [0] NCCL INFO comm 0x7ff1c0001920 rank 0 nranks 2 cudaDev 0 nvmlDev 0 - Init COMPLETE
ZBOX-EN1080K:10379:10379 [0] NCCL INFO Broadcast: opCount 0 sendbuff 0x7ff1de207e00 recvbuff 0x7ff1de207e00 count 7850 datatype 7 op 0 root 0 comm 0x7ff1c0001920 [nranks=2] stream 0x5636c1b47490
ZBOX-EN1080K:10379:10379 [0] NCCL INFO Launch mode Parallel
ZBOX-EN1080K:10379:10393 [0] NCCL INFO AllReduce: opCount 1 sendbuff 0x7ff1de207e00 recvbuff 0x7ff1de207e00 count 7850 datatype 7 op 0 root 0 comm 0x7ff1c0001920 [nranks=2] stream 0x5636c1b47490

```

rank 1

```
Namespace(backend='nccl', batch_size=128, epochs=20, init_method='tcp://30.4.241.186:23456', learning_rate=0.001, no_cuda=False, rank=1, root='data', world_size=2)
wanjia-HP-Z8-G4-Workstation:19748:19748 [0] NCCL INFO NET/Socket : Using [0]enp4s0f2:30.4.241.145<0>
wanjia-HP-Z8-G4-Workstation:19748:19748 [0] NCCL INFO NET/Plugin : No plugin found (libnccl-net.so).
wanjia-HP-Z8-G4-Workstation:19748:19748 [0] NCCL INFO NET/IB : No device found.
wanjia-HP-Z8-G4-Workstation:19748:19767 [0] NCCL INFO Setting affinity for GPU 0 to ff00ff
wanjia-HP-Z8-G4-Workstation:19748:19767 [0] NCCL INFO comm 0x7f97c4001a10 rank 1 nranks 2 cudaDev 0 nvmlDev 1
wanjia-HP-Z8-G4-Workstation:19748:19767 [0] NCCL INFO CUDA Dev 0[1], Socket NIC distance :  SOC
wanjia-HP-Z8-G4-Workstation:19748:19767 [0] NCCL INFO NET/Socket : GPU Direct RDMA Disabled for GPU 0[1] / HCA 0 (distance 3 >= 2)
wanjia-HP-Z8-G4-Workstation:19748:19767 [0] NCCL INFO Ring 00 : 0 -> 1 [receive] via NET/Socket/0
wanjia-HP-Z8-G4-Workstation:19748:19767 [0] NCCL INFO Ring 00 : 1 -> 0 [send] via NET/Socket/0
wanjia-HP-Z8-G4-Workstation:19748:19767 [0] NCCL INFO comm 0x7f97c4001a10 rank 1 nranks 2 cudaDev 0 nvmlDev 1 - Init COMPLETE
wanjia-HP-Z8-G4-Workstation:19748:19748 [0] NCCL INFO Broadcast: opCount 0 sendbuff 0x7f97cb807e00 recvbuff 0x7f97cb807e00 count 7850 datatype 7 op 0 root 0 comm 0x7f97c4001a10 [nranks=2] stream 0x5614119dff00

```

<!-- If you have a code sample, error messages, stack traces, please provide it here as well -->

## Expected behavior

it will train and print train/test loss

```
NCCL version 2.4.2+cuda10.0                                                                                                                                                                                                                                                                                                                                                     [50/1900]
ZBOX-EN1080K:10122:10131 [0] NCCL INFO Setting affinity for GPU 0 to ff
ZBOX-EN1080K:10122:10131 [0] NCCL INFO comm 0x7f0c98001920 rank 0 nranks 2 cudaDev 0 nvmlDev 0
ZBOX-EN1080K:10122:10131 [0] NCCL INFO CUDA Dev 0[0], Socket NIC distance :  PHB
ZBOX-EN1080K:10122:10131 [0] NCCL INFO Channel 00 :    0   1
ZBOX-EN1080K:10122:10131 [0] NCCL INFO Ring 00 : 1 -> 0 [receive] via NET/Socket/0
ZBOX-EN1080K:10122:10131 [0] NCCL INFO Ring 00 : 0 -> 1 [send] via NET/Socket/0
ZBOX-EN1080K:10122:10131 [0] NCCL INFO Using 256 threads, Min Comp Cap 6, Trees disabled
ZBOX-EN1080K:10122:10131 [0] NCCL INFO comm 0x7f0c98001920 rank 0 nranks 2 cudaDev 0 nvmlDev 0 - Init COMPLETE
ZBOX-EN1080K:10122:10122 [0] NCCL INFO Launch mode Parallel
Epoch: 1/20, train loss: 0.527039, train acc: 85.16%, test loss: 0.322969, test acc: 90.95%.
Epoch: 2/20, train loss: 0.319907, train acc: 91.00%, test loss: 0.291478, test acc: 91.74%.
Epoch: 3/20, train loss: 0.295930, train acc: 91.66%, test loss: 0.281328, test acc: 92.03%.
Epoch: 4/20, train loss: 0.284401, train acc: 92.05%, test loss: 0.276626, test acc: 92.12%.
Epoch: 5/20, train loss: 0.277176, train acc: 92.27%, test loss: 0.274025, test acc: 92.25%.
Epoch: 6/20, train loss: 0.272030, train acc: 92.40%, test loss: 0.272445, test acc: 92.39%.
Epoch: 7/20, train loss: 0.268088, train acc: 92.57%, test loss: 0.271436, test acc: 92.44%.
Epoch: 8/20, train loss: 0.264926, train acc: 92.64%, test loss: 0.270775, test acc: 92.48%.
Epoch: 9/20, train loss: 0.262306, train acc: 92.73%, test loss: 0.270342, test acc: 92.52%.
Epoch: 10/20, train loss: 0.260083, train acc: 92.81%, test loss: 0.270065, test acc: 92.54%.
Epoch: 11/20, train loss: 0.258161, train acc: 92.89%, test loss: 0.269901, test acc: 92.56%.
Epoch: 12/20, train loss: 0.256475, train acc: 92.91%, test loss: 0.269822, test acc: 92.61%.
Epoch: 13/20, train loss: 0.254979, train acc: 92.98%, test loss: 0.269806, test acc: 92.60%.
Epoch: 14/20, train loss: 0.253637, train acc: 93.00%, test loss: 0.269842, test acc: 92.63%.
Epoch: 15/20, train loss: 0.252423, train acc: 93.01%, test loss: 0.269919, test acc: 92.61%.
Epoch: 16/20, train loss: 0.251318, train acc: 93.04%, test loss: 0.270028, test acc: 92.60%.
Epoch: 17/20, train loss: 0.250306, train acc: 93.06%, test loss: 0.270165, test acc: 92.62%.
Epoch: 18/20, train loss: 0.249373, train acc: 93.09%, test loss: 0.270324, test acc: 92.68%.
Epoch: 19/20, train loss: 0.248510, train acc: 93.13%, test loss: 0.270501, test acc: 92.72%.
Epoch: 20/20, train loss: 0.247708, train acc: 93.14%, test loss: 0.270695, test acc: 92.71%.

```

## Environment

```
PyTorch version: 1.1.0
Is debug build: No
CUDA used to build PyTorch: 10.0.130

OS: Ubuntu 16.04.6 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.11) 5.4.0 20160609
CMake version: version 3.5.1

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration: GPU 0: GeForce GTX 1080
Nvidia driver version: 410.48
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.5.1

Versions of relevant libraries:
[pip3] numpy==1.16.4
[pip3] torch==1.1.0
[pip3] torchvision==0.2.1
[conda] blas                      1.0                         mkl    https://mirrors.ustc.edu.cn/anaconda/pkgs/main
[conda] mkl                       2019.4                      243    https://mirrors.ustc.edu.cn/anaconda/pkgs/main
[conda] mkl_fft                   1.0.12           py36ha843d7b_0    defaults
[conda] mkl_random                1.0.2            py36hd81dba3_0    defaults
[conda] pytorch                   1.1.0           py3.6_cuda10.0.130_cudnn7.5.1_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/pytorch
[conda] torchvision               0.2.1                    py36_0    defaults


## Additional context

**Comment by pietern** [comment]
There is a known problem with NCCL 2.4.2 where it locks up. You can try exporting the environment variable `NCCL_LL_THRESHOLD=0` to see if that fixes it. Otherwise, please try a newer version of PyTorch (e.g. the nightlies) that uses a newer version of NCCL where that particular bug is fixed.

You can also try running with the `gloo` backend to double check there is nothing wrong with your code.

**Comment by mrshenli** [comment]
Hi @jefby, have you got a chance to try if GLOO backend or newer version of PyTorch work for you? We did fix some deadlock errors recently.

**Comment by jefby** [comment]
Hi @mrshenli , I tried GLOO backend it's good, and i use pytorch 1.2.0.dev20190720 version with nccl this bug always exits.

**Comment by pietern** [comment]
@jefby The NCCL 2.4.8 bump happened on July 23rd. Can you try a newer nightly and post the `NCCL_DEBUG=info` output of the latest run? Perhaps it includes more information...

**Comment by GlucoseNacl** [comment]
Try setting the environment variable NCCL_P2P_DISABLE = 1
